client/cli/src/grpc_client.rs

```diff
@@ -53,80 +53,135 @@ pub struct GRPCClient {
         Ok(GRPCClient {
             client,
+            json_rpc_client,
             trusted_state: initial_trusted_state,
             latest_epoch_change_li: None,
         })
     }
+    /// Sends JSON request `request`, performs basic checks on the payload, and returns Ok(`result`),
+    /// where `result` is the payload under the key "result" in the JSON RPC response
+    /// If there is an error payload in the JSON RPC response, throw an Err with message describing the error
+    /// payload
+    fn send_json_request(&mut self, request: serde_json::Value) -> Result<serde_json::Value> {
+        let response = self
+            .json_rpc_client
+            .send_with_retry(request)?
+            .error_for_status()
+            .map_err(|e| format_err!("Server returned error: {:?}", e))?;
+
+        // check payload
+        let data: serde_json::Value = response.json()?;
+
+        // check JSON RPC protocol
+        let json_rpc_protocol = data.get("jsonrpc");
+        ensure!(
+            json_rpc_protocol == Some(&serde_json::Value::String("2.0".to_string())),
+            "JSON RPC response with incorrect protocol: {:?}",
+            json_rpc_protocol
+        );
+
+        if let Some(error) = data.get("error") {
+            bail!("Error in JSON RPC response: {:?}", error);
+        }
+
+        if let Some(result) = data.get("result") {
+            Ok(result.clone())
+        } else {
+            bail!("Received JSON RPC response with no result payload");
+        }
+    }
 
     /// Submits a transaction and bumps the sequence number for the sender, pass in `None` for
     /// sender_account if sender's address is not managed by the client.
     pub fn submit_transaction(
         &mut self,
         sender_account_opt: Option<&mut AccountData>,
-        req: &SubmitTransactionRequest,
+        transaction: SignedTransaction,
     ) -> Result<()> {
-        let mut resp = self
-            .client
-            .submit_transaction(req.clone())
-            .map_err(Into::into);
+        // form request
+        let payload = hex::encode(lcs::to_bytes(&transaction).unwrap());
+        let request = serde_json::json!({
+            "jsonrpc": "2.0",
+            "method": "submit",
+            "params": [payload],
+            "id": "Libra CLI client"
+        });
 
-        let mut try_cnt = 0;
-        while Self::need_to_retry(try_cnt, &resp) {
-            resp = self
-                .client
-                .submit_transaction(req.clone())
-                .map_err(Into::into);
-            try_cnt += 1;
-        }
-
-        let completed_resp = SubmitTransactionResponse::try_from(resp?)?;
-
-        if let Some(ac_status) = completed_resp.ac_status {
-            if ac_status == AdmissionControlStatus::Accepted {
+        match self.send_json_request(request) {
+            Ok(result) => {
+                ensure!(
+                    result == serde_json::Value::Null,
```

Comment:
nit: same, sounds it can be useful generic functionality. Not unique to transaction submission

Reply:
I think checking for result payload is Null is following the specs unique to `submit` functionality though? Each function has their own expectations for the result payload, which is why I kept this in `submit`

Reply:
oh....I totally misread that. Sorry. I thought it's error handling :)